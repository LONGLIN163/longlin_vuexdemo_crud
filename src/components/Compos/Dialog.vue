<template>
    <el-dialog title="product info" :visible="dialogFormVisible" @close='toggleDialogForm'>

        <el-form :model="form">

            <h1>{{formMode}}</h1>

            <el-form-item v-if="formMode==='edit'" label="id" :label-width="formLabelWidth">
                <el-input v-model="form.id" autocomplete="off" disabled></el-input>
            </el-form-item>

            <el-form-item label="name" :label-width="formLabelWidth">
                <el-input v-model="form.name" autocomplete="off"></el-input>
            </el-form-item>

            <el-form-item label="description" :label-width="formLabelWidth">
                <el-input v-model="form.description" autocomplete="off"></el-input>
            </el-form-item>


            <el-form-item label="taste" :label-width="formLabelWidth">
                <el-select v-model="form.taste" placeholder="please select" style="width: 100%;">
                    <el-option
                    v-for="item in types"
                    :key="item"
                    :label="item"
                    :value="item">
                    </el-option>
                </el-select>

            </el-form-item>


            <el-form-item label="color" :label-width="formLabelWidth">
                <div class="colorBox">
                    <el-color-picker v-model="form.color"></el-color-picker>
                    <div class="colorTxt">{{form.color}}</div>
                </div>
            </el-form-item>

            <el-form-item label="image" :label-width="formLabelWidth">
                <el-input v-model="form.image" autocomplete="off"></el-input>
            </el-form-item>


            <el-form-item label="expires" :label-width="formLabelWidth">
                
                    <el-date-picker
                        v-model="form.expires"
                        type="date"
                        style="width: 100%;"
                        placeholder="choose a date"
                    >
                    </el-date-picker>
                
            </el-form-item>

            <el-form-item label="price" :label-width="formLabelWidth">
                <el-input v-model="form.price" autocomplete="off"></el-input>
            </el-form-item>

        </el-form>

        <div slot="footer" class="dialog-footer">
            <el-button @click="toggleDialogForm">Cancel</el-button>
            <el-button v-show="formMode===''" type="primary" @click="handleCreate(form)">Create</el-button>
        </div>

    </el-dialog>
</template>

<script>
    import store  from '@/vuex/store'
    import {mapState,mapMutations,mapActions} from 'vuex'

    export default {
        data(){
            return {
                formLabelWidth: '120px',
            }
        },
        props:['formMode','form'],
        computed:{
          ...mapState(['articleList','dialogFormVisible','types'])
        },
        methods: {
          handleCreate(form){
            this.createArticleAction(form);
            this.toggleDialogForm()
          },
          ...mapMutations(['toggleDialogForm']),
          ...mapActions(['createArticleAction'])
        },
        store
    }
</script>

<style scoped>
    .colorTxt{
        height: 40px;
        line-height: 40px;
        margin-left: 10px;
    }
    .colorBox{
        display: flex;
        flex-direction: row;
        align-items: flex-start
    }

</style>