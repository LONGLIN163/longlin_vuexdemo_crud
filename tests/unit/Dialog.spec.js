import { mount,createLocalVue} from '@vue/test-utils'
import Dialog from '@/components/Compos/Dialog'
import {mockTypes} from '../mockedData'

import Vuex from 'vuex'
import ElementUI from "element-ui"
import locale from 'element-ui/lib/locale/lang/en'

const localVue = createLocalVue();
localVue.use(ElementUI,{locale});  // install ElementUI to the local vue instance
localVue.use(Vuex); // install vuex to the local vue instance

let wrapper;
describe('Component Dialog tests:', () => {

  // define them outside of breforeEach
  let actions
  let state
  let store
  let mutations
  let handleCreate
  
  beforeEach(() => {

    // mock handleCreate()
    handleCreate = jest.spyOn(Dialog.methods, 'handleCreate').mockImplementation(() => {
      // we don't need have to mock mutations here
      actions.createArticleAction()
      mutations.TOOGLE_DIALOGFORM()
    })

    // mock actions
    actions = {
      createArticleAction: jest.fn()
    }

    // mock mutations
    mutations = {
      TOOGLE_DIALOGFORM: jest.fn()
    }

    //mock state
    state={
      dialogFormVisible:true,
      types:mockTypes
    }

    // create a fake store
    store = new Vuex.Store({
      state,
      actions,
      mutations
    })

    wrapper = mount(Dialog,{
      localVue,
      store,
      propsData:{
        formMode:'', 
        form: {
            name: '',
            description: '',
            taste: '',
            color: '',
            image: '',
            isFruit: true,
            expires: '',
            price: ''
        },
         
      }
    })
  })
  
  
  it('Get Dialog snapshot', async () => {

    expect(wrapper).toMatchSnapshot()

  })
    
  it('test if handleCreate and createArticleAction have been called',async () => {

    const createBtn=await wrapper.find('#createArticleBtn');
    await createBtn.trigger('click')
    expect(handleCreate).toHaveBeenCalled()
    expect(actions.createArticleAction).toHaveBeenCalled()
    expect(mutations.TOOGLE_DIALOGFORM).toHaveBeenCalled()

  })

  it('test if TOOGLE_DIALOGFORM has been called when click cancel',async () => {

    const cancelCreateBtn=await wrapper.find('#cancelCreateBtn');
    await cancelCreateBtn.trigger('click')
    expect(mutations.TOOGLE_DIALOGFORM).toHaveBeenCalled()

  })

  it('test form elements',async () => {
    
    //test-input name field
    const nameInput=wrapper.find('[data-test="dialog_name"]');
    await nameInput.setValue("haha")
    expect(wrapper.find('[data-test="dialog_name"]').element.value).toBe('haha')

    //test-input description field
    const descriptionInput=wrapper.find('[data-test="dialog_description"]');
    await descriptionInput.setValue("xixi")
    expect(wrapper.find('[data-test="dialog_description"]').element.value).toBe('xixi')

    //test-type select field
    //const typeSelect =await wrapper.find('[data-test="taste"] .el-select .el-input__inner').element.placeholder
    //const typeSelect = wrapper.find('[data-test="taste"]>div>div>div>input').element.placeholder
    const typeSelect = wrapper.find('[data-test="taste"]>div>div>div>input')
    await typeSelect.trigger("click")
    const typeSelectItem = wrapper.findAll('[data-test="dialog_type"]').at(0)
    expect(typeSelectItem.text()).toBe('Refined')
    await typeSelectItem.trigger("click")
    expect(typeSelect.element.value).toBe('Refined')
    
    //test- picker field
    // const colorSelect = wrapper.find('[data-test="dialog_color"]>div>div>div')
    // await colorSelect.trigger("click")
    // const colorSelectInput = wrapper.find('.el-color-picker__panel .el-color-dropdown__value .el-input__inner')
    // await colorSelectInput.setValue('#B86C6C')
    // expect(colorSelectInput.element.value).toBe('#B86C6C')
    // wrapper.findAll('.el-button--mini').at(1).trigger("click").then(() => {
    //   // ***there is a bug, asyn+await would not work here, why???***
    //   const colorSelectedTxt = wrapper.find('[data-test="colorTxt"]')
    //   console.log("******",colorSelectedTxt.text())
    //   expect(colorSelectedTxt.text()).toBe('#B86C6Csss')
    // })

    //test-date picker field
    // const dateSelect = wrapper.find('[data-test="dialog_date"]>div>div')
    // dateSelect.trigger("click").then(async () => {
    //   const dateSelectCell = await wrapper.find('.el-date-picker tbody tr:first-child td:first-child')
    //   await dateSelectCell.trigger("click")
    //   console.log("******",dateSelect.text())
    // })

    /**
     * date picker and color picker will be tested in e2e
     * skip the rest form tests
     */


  })

  afterEach(() => {
    wrapper.destroy();
  });


})