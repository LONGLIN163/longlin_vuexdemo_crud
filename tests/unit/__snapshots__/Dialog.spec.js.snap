// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Dialog tests: Get Dialog snapshot 1`] = `
<transition-stub name="dialog-fade" style="z-index: 2001;">
  <div class="el-dialog__wrapper">
    <div role="dialog" aria-modal="true" aria-label="dialog" class="el-dialog" style="margin-top: 15vh;">
      <div class="el-dialog__header"><span class="el-dialog__title"></span><button type="button" aria-label="Close" class="el-dialog__headerbtn"><i class="el-dialog__close el-icon el-icon-close"></i></button></div>
      <div class="el-dialog__body">
        <form class="el-form">
          <h1 id="productinfo">Product info</h1>
          <div class="el-form-item"><label class="el-form-item__label" style="width: 120px;">name</label>
            <div class="el-form-item__content" style="margin-left: 120px;">
              <div class="el-input">
                <!----><input type="text" autocomplete="off" data-test="dialog_name" id="aname" class="el-input__inner">
                <!---->
                <!---->
                <!---->
                <!---->
              </div>
              <transition-stub name="el-zoom-in-top">
                <!---->
              </transition-stub>
            </div>
          </div>
          <div class="el-form-item"><label class="el-form-item__label" style="width: 120px;">description</label>
            <div class="el-form-item__content" style="margin-left: 120px;">
              <div class="el-input">
                <!----><input type="text" autocomplete="off" data-test="dialog_description" class="el-input__inner">
                <!---->
                <!---->
                <!---->
                <!---->
              </div>
              <transition-stub name="el-zoom-in-top">
                <!---->
              </transition-stub>
            </div>
          </div>
          <div class="el-form-item" data-test="taste"><label class="el-form-item__label" style="width: 120px;">taste</label>
            <div class="el-form-item__content" style="margin-left: 120px;">
              <div class="el-select" style="width: 100%;">
                <!---->
                <div class="el-input el-input--suffix">
                  <!----><input type="text" readonly="readonly" autocomplete="off" placeholder="please select" class="el-input__inner">
                  <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----><!----><!----><!----></span>
                  <!----></span>
                  <!---->
                  <!---->
                </div>
                <transition-stub name="el-zoom-in-top">
                  <div class="el-select-dropdown el-popper" style="display: none;">
                    <div class="el-scrollbar" style="">
                      <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
                        <ul class="el-scrollbar__view el-select-dropdown__list">
                          <!---->
                          <li class="el-select-dropdown__item" data-test="dialog_type"><span>Refined</span></li>
                          <li class="el-select-dropdown__item" data-test="dialog_type"><span>Unbranded</span></li>
                          <li class="el-select-dropdown__item" data-test="dialog_type"><span>Licensed</span></li>
                          <li class="el-select-dropdown__item" data-test="dialog_type"><span>Small</span></li>
                          <li class="el-select-dropdown__item" data-test="dialog_type"><span>Sleek</span></li>
                          <li class="el-select-dropdown__item" data-test="dialog_type"><span>Awesome</span></li>
                        </ul>
                      </div>
                      <div class="el-scrollbar__bar is-horizontal">
                        <div class="el-scrollbar__thumb" style="transform: translateX(0%); webkit-transform: translateX(0%);"></div>
                      </div>
                      <div class="el-scrollbar__bar is-vertical">
                        <div class="el-scrollbar__thumb" style="transform: translateY(0%); webkit-transform: translateY(0%);"></div>
                      </div>
                    </div>
                    <!---->
                  </div>
                </transition-stub>
              </div>
              <transition-stub name="el-zoom-in-top">
                <!---->
              </transition-stub>
            </div>
          </div>
          <div class="el-form-item" data-test="dialog_color"><label class="el-form-item__label" style="width: 120px;">color</label>
            <div class="el-form-item__content" style="margin-left: 120px;">
              <div class="colorBox">
                <div class="el-color-picker">
                  <!---->
                  <div class="el-color-picker__trigger"><span class="el-color-picker__color"><span class="el-color-picker__color-inner" style="background-color: transparent;"></span><span class="el-color-picker__empty el-icon-close"></span></span><span class="el-color-picker__icon el-icon-arrow-down" style="display: none;"></span></div>
                  <transition-stub name="el-zoom-in-top" class="el-color-picker__panel">
                    <div class="el-color-dropdown" style="display: none;">
                      <div class="el-color-dropdown__main-wrapper">
                        <div class="el-color-hue-slider is-vertical" style="float: right;">
                          <div class="el-color-hue-slider__bar"></div>
                          <div class="el-color-hue-slider__thumb" style="left: 0px; top: 0px;"></div>
                        </div>
                        <div class="el-color-svpanel" style="background-color: rgb(255, 0, 0);">
                          <div class="el-color-svpanel__white"></div>
                          <div class="el-color-svpanel__black"></div>
                          <div class="el-color-svpanel__cursor" style="top: 0px; left: 0px;">
                            <div></div>
                          </div>
                        </div>
                      </div>
                      <!---->
                      <!---->
                      <div class="el-color-dropdown__btns"><span class="el-color-dropdown__value"><div class="el-input el-input--mini"><!----><input type="text" autocomplete="off" class="el-input__inner"><!----><!----><!----><!----></div></span><button type="button" class="el-button el-color-dropdown__link-btn el-button--text el-button--mini">
                          <!---->
                          <!----><span>
        Clear
      </span>
                        </button><button type="button" class="el-button el-color-dropdown__btn el-button--default el-button--mini is-plain">
                          <!---->
                          <!----><span>
        OK
      </span>
                        </button></div>
                    </div>
                  </transition-stub>
                </div>
                <div data-test="colorTxt" class="colorTxt"></div>
              </div>
              <transition-stub name="el-zoom-in-top">
                <!---->
              </transition-stub>
            </div>
          </div>
          <div class="el-form-item"><label class="el-form-item__label" style="width: 120px;">image</label>
            <div class="el-form-item__content" style="margin-left: 120px;">
              <div class="el-input">
                <!----><input type="text" autocomplete="off" class="el-input__inner">
                <!---->
                <!---->
                <!---->
                <!---->
              </div>
              <transition-stub name="el-zoom-in-top">
                <!---->
              </transition-stub>
            </div>
          </div>
          <div class="el-form-item" data-test="dialog_date"><label class="el-form-item__label" style="width: 120px;">expires</label>
            <div class="el-form-item__content" style="margin-left: 120px;">
              <div class="el-date-editor el-input el-input--prefix el-input--suffix el-date-editor--date" style="width: 100%;">
                <!----><input type="text" autocomplete="off" name="" placeholder="choose a date" class="el-input__inner"><span class="el-input__prefix"><i class="el-input__icon el-icon-date"></i><!----></span><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-input__icon"></i><!----><!----><!----><!----></span>
                <!----></span>
                <!---->
                <!---->
              </div>
              <transition-stub name="el-zoom-in-top">
                <!---->
              </transition-stub>
            </div>
          </div>
          <div class="el-form-item"><label class="el-form-item__label" style="width: 120px;">price</label>
            <div class="el-form-item__content" style="margin-left: 120px;">
              <div class="el-input">
                <!----><input type="text" autocomplete="off" class="el-input__inner">
                <!---->
                <!---->
                <!---->
                <!---->
              </div>
              <transition-stub name="el-zoom-in-top">
                <!---->
              </transition-stub>
            </div>
          </div>
        </form>
      </div>
      <div class="el-dialog__footer">
        <div class="dialog-footer"><button type="button" class="el-button el-button--default" id="cancelCreateBtn">
            <!---->
            <!----><span>Cancel</span>
          </button> <button type="button" class="el-button el-button--primary" id="createArticleBtn">
            <!---->
            <!----><span>Create</span>
          </button></div>
      </div>
    </div>
  </div>
</transition-stub>
`;
